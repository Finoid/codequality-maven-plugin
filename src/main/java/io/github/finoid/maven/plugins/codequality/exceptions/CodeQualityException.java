package io.github.finoid.maven.plugins.codequality.exceptions;

public class CodeQualityException extends RuntimeException {
    public CodeQualityException(final String message) {
        super(message);
    }

    public CodeQualityException(final String message, final Throwable cause) {
        super(message, cause);
    }
}
