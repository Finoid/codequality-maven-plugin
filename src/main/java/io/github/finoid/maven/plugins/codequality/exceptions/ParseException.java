package io.github.finoid.maven.plugins.codequality.exceptions;

public class ParseException extends CodeQualityException {
    public ParseException(final String message, final Throwable cause) {
        super(message, cause);
    }
}
