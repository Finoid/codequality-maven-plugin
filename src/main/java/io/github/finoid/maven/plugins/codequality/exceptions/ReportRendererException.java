package io.github.finoid.maven.plugins.codequality.exceptions;

public class ReportRendererException extends CodeQualityException {
    public ReportRendererException(final String message, final Throwable cause) {
        super(message, cause);
    }
}
