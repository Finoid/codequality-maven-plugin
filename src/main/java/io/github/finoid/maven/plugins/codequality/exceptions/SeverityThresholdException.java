package io.github.finoid.maven.plugins.codequality.exceptions;

public class SeverityThresholdException extends CodeQualityException {
    public SeverityThresholdException(final String message) {
        super(message);
    }
}
