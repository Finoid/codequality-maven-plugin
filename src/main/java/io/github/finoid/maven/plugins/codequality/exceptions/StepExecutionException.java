package io.github.finoid.maven.plugins.codequality.exceptions;

public class StepExecutionException extends CodeQualityException {
    public StepExecutionException(final String message, final Throwable cause) {
        super(message, cause);
    }
}
