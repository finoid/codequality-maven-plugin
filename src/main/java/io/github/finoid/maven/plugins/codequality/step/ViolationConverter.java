package io.github.finoid.maven.plugins.codequality.step;

import com.puppycrawl.tools.checkstyle.api.AuditEvent;
import com.puppycrawl.tools.checkstyle.api.SeverityLevel;
import io.github.finoid.maven.plugins.codequality.report.Severity;
import io.github.finoid.maven.plugins.codequality.report.Violation;
import io.github.finoid.maven.plugins.codequality.util.Precondition;
import lombok.SneakyThrows;
import org.apache.maven.project.MavenProject;

import javax.inject.Inject;
import javax.inject.Singleton;
import javax.xml.bind.DatatypeConverter;
import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.util.Locale;
import java.util.regex.Matcher;

@Singleton
public class ViolationConverter {
    private final MavenProject project;

    @Inject
    public ViolationConverter(final MavenProject project) {
        this.project = Precondition.nonNull(project, "MavenProject shouldn't be null");
    }

    public Violation ofAuditEvent(final AuditEvent auditEvent) {
        final File repositoryRoot = project.getBasedir();

        return new Violation(
            String.format("%s: %s", "Checkstyle", auditEvent.getMessage()),
            fingerprint(repositoryRoot, auditEvent),
            severity(auditEvent.getSeverityLevel()),
            relativePath(repositoryRoot, auditEvent.getFileName()),
            auditEvent.getFileName().replace("\\", "/"),  // Windows compatibility
            lineNumber(auditEvent),
            auditEvent.getColumn(),
            auditEvent.getViolation().getKey() + " - " + sourceName(auditEvent)
        );
    }

    public Violation ofErrorProneViolationMatcher(final Matcher violationMatcher) {
        final File repositoryRoot = project.getBasedir();

        final String columnNumber = violationMatcher.group("column");
        final String absoluteFilePath = violationMatcher.group("path");
        final String description = violationMatcher.group("description");
        final int lineNumber = Integer.parseInt(violationMatcher.group("line"));
        final String rule = violationMatcher.group("rule");

        final int column = columnNumber == null ? 0 : Integer.parseInt(columnNumber);

        return new Violation(
            String.format("%s: %s", "ErrorProne", description),
            fingerprint(repositoryRoot, absoluteFilePath, description, lineNumber, columnNumber),
            Severity.MINOR,
            relativePath(repositoryRoot, absoluteFilePath),
            absoluteFilePath.replace("\\", "/"),  // Windows compatibility
            lineNumber,
            column,
            rule);
    }

    public Violation ofCheckerFrameworkViolationMatcher(final Matcher violationMatcher) {
        final File repositoryRoot = project.getBasedir();

        final String columnNumber = violationMatcher.group("column");
        final String absoluteFilePath = violationMatcher.group("path");
        final String description = violationMatcher.group("description");
        final int lineNumber = Integer.parseInt(violationMatcher.group("line"));
        final String rule = violationMatcher.group("rule");

        return new Violation(
            String.format("%s: %s", "CheckerFramework", description),
            fingerprint(repositoryRoot, absoluteFilePath, description, lineNumber, columnNumber),
            Severity.MINOR,
            relativePath(repositoryRoot, absoluteFilePath),
            absoluteFilePath.replace("\\", "/"), // Windows compatibility
            lineNumber,
            Integer.valueOf(columnNumber),
            rule);
    }

    @SneakyThrows
    private String fingerprint(final File repositoryRoot, final AuditEvent auditEvent) {
        final String key = String.format("%s:%s:%s:%d",
            relativePath(repositoryRoot, auditEvent.getFileName()),
            auditEvent.getSeverityLevel(),
            auditEvent.getMessage(),
            auditEvent.getLine() + auditEvent.getColumn()
        );

        return fingerprint(key);
    }

    @SneakyThrows
    private String fingerprint(final File repositoryRoot, final String path, final String message, final int lineNumber, final String column) {
        final String key = String.format("%s:%s:%s:%s",
            relativePath(repositoryRoot, path),
            "WARNING",
            message,
            lineNumber + column
        );

        return fingerprint(key);
    }

    @SneakyThrows
    private String fingerprint(final String content) {
        final MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
        messageDigest.update(content.getBytes(StandardCharsets.UTF_8));

        final byte[] digest = messageDigest.digest();

        return DatatypeConverter.printHexBinary(digest)
            .toLowerCase(Locale.ROOT);
    }

    private String relativePath(final File repositoryRoot, final String absoluteFilePath) {
        final Path absolutePath = Path.of(absoluteFilePath);

        return repositoryRoot.toPath()
            .relativize(absolutePath)
            .toString()
            .replace("\\", "/"); // Windows compatibility
    }

    private static String sourceName(final AuditEvent auditEvent) {
        String sourceName = auditEvent.getSourceName();

        if (sourceName.endsWith("Check")) {
            sourceName = sourceName.substring(0, sourceName.length() - 5);
        }

        return sourceName.substring(sourceName.lastIndexOf(".") + 1);
    }

    private static int lineNumber(final AuditEvent auditEvent) {
        return auditEvent.getLine();
    }

    private Severity severity(final SeverityLevel severity) {
        switch (severity) {
            case ERROR:
                return Severity.MAJOR;
            case WARNING:
                return Severity.MINOR;
            default:
                return Severity.INFO;
        }
    }
}
