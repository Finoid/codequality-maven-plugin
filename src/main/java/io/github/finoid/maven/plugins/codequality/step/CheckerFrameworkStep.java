package io.github.finoid.maven.plugins.codequality.step;

import io.github.finoid.maven.plugins.codequality.MavenAnnotationProcessorsManager;
import io.github.finoid.maven.plugins.codequality.configuration.CheckerFrameworkConfiguration;
import io.github.finoid.maven.plugins.codequality.configuration.CodeQualityConfiguration;
import io.github.finoid.maven.plugins.codequality.exceptions.CodeQualityException;
import io.github.finoid.maven.plugins.codequality.report.CheckerFrameworkViolationLogParser;
import io.github.finoid.maven.plugins.codequality.report.Violation;
import io.github.finoid.maven.plugins.codequality.util.CollectorUtils;
import io.github.finoid.maven.plugins.codequality.util.MojoUtils.ElementUtils;
import io.github.finoid.maven.plugins.codequality.util.MojoUtils.PluginUtils;
import io.github.finoid.maven.plugins.codequality.util.Precondition;
import io.github.finoid.maven.plugins.codequality.util.ProjectUtils;
import io.github.finoid.maven.plugins.codequality.util.PropertyUtils;
import org.apache.maven.execution.MavenSession;
import org.apache.maven.plugin.BuildPluginManager;
import org.apache.maven.plugin.descriptor.PluginDescriptor;
import org.apache.maven.plugin.logging.Log;
import org.apache.maven.project.MavenProject;
import org.twdata.maven.mojoexecutor.MojoExecutor;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import static org.twdata.maven.mojoexecutor.MojoExecutor.configuration;
import static org.twdata.maven.mojoexecutor.MojoExecutor.element;
import static org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo;
import static org.twdata.maven.mojoexecutor.MojoExecutor.executionEnvironment;
import static org.twdata.maven.mojoexecutor.MojoExecutor.goal;

/**
 * Step which executes the CheckerFrameworkStep analyzer.
 */
@Singleton
public class CheckerFrameworkStep implements Step<CheckerFrameworkConfiguration> {
    private final MavenProject project;
    private final MavenSession mavenSession;
    private final BuildPluginManager pluginManager;
    private final CheckerFrameworkViolationLogParser checkerFrameworkViolationLogParser;

    @Inject
    public CheckerFrameworkStep(
        final MavenProject project,
        final MavenSession mavenSession,
        final BuildPluginManager pluginManager,
        final CheckerFrameworkViolationLogParser checkerFrameworkViolationLogParser
    ) {
        this.project = Precondition.nonNull(project, "MavenProject shouldn't be null");
        this.mavenSession = Precondition.nonNull(mavenSession, "MavenSession shouldn't be null");
        this.pluginManager = Precondition.nonNull(pluginManager, "BuildPluginManager shouldn't be null");
        this.checkerFrameworkViolationLogParser =
            Precondition.nonNull(checkerFrameworkViolationLogParser, "CheckerFrameworkViolationLogParser shouldn't be null");
    }

    @Override
    public boolean isEnabled(final CheckerFrameworkConfiguration configuration) {
        return configuration.isEnabled();
    }

    @Override
    public PrerequisiteResult hasPrerequisites(final CheckerFrameworkConfiguration configuration) {
        if (ProjectUtils.isPresentOnClassPath(mavenSession.getCurrentProject(), "org.checkerframework", "checker-qual")) {
            return PrerequisiteResult.OK;
        }

        return PrerequisiteResult.notOK("org.checkerframework.checker-qual is missing on the class path");
    }

    @Override
    public StepType type() {
        return StepType.CHECKER_FRAMEWORK;
    }

    @Override
    public StepResult execute(final CodeQualityConfiguration codeQualityConfiguration, final CheckerFrameworkConfiguration checkerFrameworkConfiguration,
                              final Log log) {
        final List<Violation> violations = executeStep(codeQualityConfiguration, checkerFrameworkConfiguration, log);

        return StepResult.create(StepType.CHECKER_FRAMEWORK, checkerFrameworkConfiguration.isPermissive(), violations);
    }

    @Override
    public CleanContext getCleanContext() {
        return CleanContext.DO_NOTHING;
    }

    private List<Violation> executeStep(
        final CodeQualityConfiguration codeQualityConfiguration,
        final CheckerFrameworkConfiguration stepConfiguration,
        final Log log
    ) {
        final PluginDescriptor descriptor =
            PluginUtils.pluginDescriptor("org.apache.maven.plugins", "maven-compiler-plugin", codeQualityConfiguration.getVersions().getMavenCompiler());

        final MavenProject currentProject = mavenSession.getCurrentProject();

        final File currentProjectArtifactFile = currentProject.getArtifact()
            .getFile();

        final String javaVersion = PropertyUtils.valueOrFallback(project.getProperties(), "java.version", "21");

        try {
            executeMojo(
                PluginUtils.pluginOfDescriptor(descriptor),
                goal("compile"),
                configuration(
                    element(MojoExecutor.name("fork"), "true"), // To be able to apply javac flags, see elementsOfCompilerArgs
                    element("source", javaVersion),
                    element("target", javaVersion),
                    element("release", javaVersion),
                    element("outputDirectory", project.getBuild().getDirectory() + "/checker-framework-classes"),
                    element("failOnError", "false"),
                    element("showWarnings", "true"),
                    element(MojoExecutor.name("compilerArgs"), elementsOfCompilerArgs(stepConfiguration)
                        .toArray(MojoExecutor.Element[]::new)),
                    element(MojoExecutor.name("annotationProcessorPaths"),
                        elementsOfAnnotationProcessorPaths(currentProject, codeQualityConfiguration, stepConfiguration)
                            .toArray(MojoExecutor.Element[]::new)),
                    element(MojoExecutor.name("annotationProcessors"),
                        elementsOfCheckers(currentProject, stepConfiguration.getCheckers(), codeQualityConfiguration)
                            .toArray(MojoExecutor.Element[]::new))
                ),
                executionEnvironment(currentProject, mavenSession, pluginManager)
            );

            // Restores the current project's original artifact file, resolving
            // the error: 'The packaging for this project did not assign a file to the build artifact.'
            currentProject.getArtifact()
                .setFile(currentProjectArtifactFile);

            return parseViolations(log);
        } catch (final Exception e) {
            throw new CodeQualityException("Error during execution of CheckerFramework step", e);
        }
    }

    private List<MojoExecutor.Element> elementsOfCompilerArgs(final CheckerFrameworkConfiguration checkerFrameworkConfiguration) {
        final List<MojoExecutor.Element> compilerArgs = checkerFrameworkConfiguration.getCompilerArgs().stream()
            .map(compilerArg -> element(MojoExecutor.name("arg"), compilerArg))
            .collect(CollectorUtils.toMutableList());

        // Due to JEP 396: Strongly Encapsulate JDK Internals by Default - See https://checkerframework.org/manual/checker-framework-manual.pdf
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED"));
        compilerArgs.add(element(MojoExecutor.name("arg"), "-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED"));

        // Checker framework rules that are suppressed by default
        compilerArgs.add(
            element(MojoExecutor.name("arg"), "-AsuppressWarnings=type.anno.before.decl.anno,type.anno.before.modifier")); // TODO (nw) should be configurable

        // Output errors as warnings
        compilerArgs.add(element(MojoExecutor.name("arg"), "-Awarns"));

        // The -processing suppress "No processor claimed any of these annotations"
        // Suppress warnings related to JPMS due to compatibility issues with lombok
        compilerArgs.add(element(MojoExecutor.name("arg"),
            "-Xlint:all,-serial,-processing,-requires-transitive-automatic,-missing-explicit-ctor,-exports,-requires-automatic"));

        // Skip target directory which includes generated sources
        compilerArgs.add(element(MojoExecutor.name("arg"), "-AskipFiles=/target/"));

        return compilerArgs;
    }

    private List<MojoExecutor.Element> elementsOfAnnotationProcessorPaths(final MavenProject currentProject,
                                                                          final CodeQualityConfiguration codeQualityConfiguration,
                                                                          final CheckerFrameworkConfiguration checkerFrameworkConfiguration) {
        final MavenAnnotationProcessorsManager annotationProcessorsManager = new MavenAnnotationProcessorsManager(currentProject, codeQualityConfiguration);

        final List<MojoExecutor.Element> annotationProcessorPaths = annotationProcessorsManager.annotationPaths().stream()
            .map(it -> ElementUtils.annotationProcessor(it.getGroupId(), it.getArtifactId(), it.getVersion()))
            .collect(CollectorUtils.toMutableList());

        annotationProcessorPaths.add(
            ElementUtils.annotationProcessor("org.checkerframework", "checker", checkerFrameworkConfiguration.getVersions().getCheckerFramework()));

        return annotationProcessorPaths;
    }

    private List<MojoExecutor.Element> elementsOfCheckers(final MavenProject currentProject, final Set<String> checkers,
                                                          final CodeQualityConfiguration codeQualityConfiguration) {
        final MavenAnnotationProcessorsManager annotationProcessorsManager = new MavenAnnotationProcessorsManager(currentProject, codeQualityConfiguration);

        return annotationProcessorsManager.annotationProcessors(checkers).stream()
            .map(annotationProcessor -> element(MojoExecutor.name("annotationProcessor"), annotationProcessor))
            .toList();
    }

    private List<Violation> parseViolations(final Log log) {
        final String checkerFrameworkOutputFilePath = checkerFrameworkOutputFilePath(project);

        return violationsFromOutputFile(checkerFrameworkOutputFilePath, log);
    }

    private List<Violation> violationsFromOutputFile(final String checkerFrameworkOutputFilePath, final Log log) {
        try (final InputStream targetStream = new FileInputStream(checkerFrameworkOutputFilePath)) {
            return checkerFrameworkViolationLogParser.parse(targetStream);
        } catch (final IOException e) {
            log.warn("No checker framework file found. Please register the plugin as a extension");

            return Collections.emptyList();
        }
    }

    private String checkerFrameworkOutputFilePath(final MavenProject project) {
        return targetOutputFilePath(project.getBuild().getDirectory(),
            String.format("checkerframework-%s.txt", mavenSession.getCurrentProject().getModel().getArtifactId()));
    }

    private static String targetOutputFilePath(final String targetDirectory, final String targetOutputFilename) {
        return targetDirectory + "/" + targetOutputFilename;
    }
}
