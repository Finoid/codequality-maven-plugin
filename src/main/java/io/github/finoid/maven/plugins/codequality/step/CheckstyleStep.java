package io.github.finoid.maven.plugins.codequality.step;

import io.github.finoid.maven.plugins.codequality.configuration.CheckstyleConfiguration;
import io.github.finoid.maven.plugins.codequality.configuration.CodeQualityConfiguration;
import io.github.finoid.maven.plugins.codequality.exceptions.CodeQualityException;
import io.github.finoid.maven.plugins.codequality.factories.CheckstyleExecutorRequestFactory;
import io.github.finoid.maven.plugins.codequality.report.Violation;
import io.github.finoid.maven.plugins.codequality.util.Precondition;
import lombok.SneakyThrows;
import org.apache.maven.plugin.logging.Log;
import org.apache.maven.plugins.checkstyle.exec.CheckstyleExecutor;
import org.apache.maven.plugins.checkstyle.exec.CheckstyleExecutorRequest;
import org.apache.maven.plugins.checkstyle.exec.CheckstyleResults;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.Collections;
import java.util.List;

/**
 * Step which executes the Checkstyle analyzer.
 */
@Singleton
public class CheckstyleStep implements Step<CheckstyleConfiguration> {
    private final CheckstyleExecutor checkstyleExecutor;
    private final CheckstyleExecutorRequestFactory checkstyleExecutorRequestFactory;
    private final ViolationConverter violationConverter;

    @Inject
    public CheckstyleStep(
        final CheckstyleExecutor checkstyleExecutor,
        final CheckstyleExecutorRequestFactory checkstyleExecutorRequestFactory,
        final ViolationConverter violationConverter
    ) {
        this.checkstyleExecutor = Precondition.nonNull(checkstyleExecutor, "CheckstyleExecutor shouldn't be null");
        this.checkstyleExecutorRequestFactory = Precondition.nonNull(checkstyleExecutorRequestFactory, "CheckstyleExecutorRequestFactory shouldn't be null");
        this.violationConverter = Precondition.nonNull(violationConverter, "ViolationConverter shouldn't be null");
    }

    @Override
    public boolean isEnabled(final CheckstyleConfiguration configuration) {
        return configuration.isEnabled();
    }

    @Override
    public StepType type() {
        return StepType.CHECKSTYLE;
    }

    @Override
    public StepResult execute(final CodeQualityConfiguration codeQualityConfiguration, final CheckstyleConfiguration stepConfiguration, final Log log) {
        try {
            final StepResult resultMain = executeForEnvironment(stepConfiguration, stepConfiguration.getExecutionMain(), log);
            final StepResult resultTest = executeForEnvironment(stepConfiguration, stepConfiguration.getExecutionTest(), log);

            return StepResult.create(StepType.CHECKSTYLE, stepConfiguration.isPermissive(), resultMain.getViolations(), resultTest.getViolations());
        } catch (final Exception e) {
            throw new CodeQualityException("Error during execution of checkstyle step", e);
        }
    }

    @Override
    public CleanContext getCleanContext() {
        return CleanContext.DO_NOTHING;
    }

    private StepResult executeForEnvironment(
        final CheckstyleConfiguration configuration,
        final CheckstyleConfiguration.ExecutionEnvironment executionEnvironment,
        final Log log
    ) {
        return executeCheckstyle(configuration, executionEnvironment, log);
    }

    @SneakyThrows
    private StepResult executeCheckstyle(final CheckstyleConfiguration configuration, final CheckstyleConfiguration.ExecutionEnvironment executionEnvironment,
                                         final Log log) {
        if (!executionEnvironment.isEnabled()) {
            log.info("Skipping Checkstyle Sub Step for " + executionEnvironment);

            return StepResult.create(StepType.CHECKSTYLE, configuration.isPermissive(), Collections.emptyList());
        }

        log.info("Executing Checkstyle Sub Step for " + executionEnvironment.getEnvironment());

        final CheckstyleExecutorRequest request = checkstyleExecutorRequestFactory.create(configuration, executionEnvironment, log);

        final CheckstyleResults checkstyleResults = checkstyleExecutor.executeCheckstyle(request);

        final List<Violation> violations = checkstyleResults.getFiles()
            .entrySet()
            .stream()
            .flatMap(it -> it.getValue().stream())
            .map(violationConverter::ofAuditEvent)
            .toList();

        return StepResult.create(StepType.CHECKSTYLE, configuration.isPermissive(), violations);
    }
}
