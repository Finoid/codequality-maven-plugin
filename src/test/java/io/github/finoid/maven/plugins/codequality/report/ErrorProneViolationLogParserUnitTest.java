package io.github.finoid.maven.plugins.codequality.report;

import io.github.finoid.maven.plugins.codequality.exceptions.ParseException;
import io.github.finoid.maven.plugins.codequality.fixtures.ResourceUtils;
import io.github.finoid.maven.plugins.codequality.fixtures.TemplateResourceUtils;
import io.github.finoid.maven.plugins.codequality.fixtures.UnitTest;
import io.github.finoid.maven.plugins.codequality.step.ViolationConverter;
import org.apache.maven.project.MavenProject;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.Mockito;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.nio.file.Paths;

class ErrorProneViolationLogParserUnitTest extends UnitTest {
    @Mock
    private MavenProject project;

    private ErrorProneViolationLogParser unit;

    @BeforeEach
    void beforeEach() {
        unit = new ErrorProneViolationLogParser(new ViolationConverter(project));
    }

    @Test
    void givenValidLogFile_whenParse_thenExpectedViolations() {
        var logFileInputStream = ResourceUtils.tryInputStreamFrom("files/errorprone-logs.template.txt");

        var projectRootPath = Paths.get("")
            .toAbsolutePath();

        var content = TemplateResourceUtils.template(logFileInputStream, projectRootPath.toString());

        Mockito.when(project.getBasedir())
            .thenReturn(projectRootPath.toFile());

        var result = unit.parse(new ByteArrayInputStream(content.getBytes()));

        snapshot(result, "*..fullPath");
    }

    @Test
    void givenClosedInputStream_whenParse_thenExpectedException() throws IOException {
        var logFileInputStream = ResourceUtils.tryInputStreamFrom("files/errorprone-logs.template.txt");
        logFileInputStream.close();

        Assertions.assertThrows(ParseException.class, () -> unit.parse(logFileInputStream));
    }

}