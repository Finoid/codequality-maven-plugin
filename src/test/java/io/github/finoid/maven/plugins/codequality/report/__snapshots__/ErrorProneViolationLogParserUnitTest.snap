io.github.finoid.maven.plugins.codequality.report.ErrorProneViolationLogParserUnitTest.givenValidLogFile_whenParse_thenExpectedViolations=[
  [
    {
      "columnNumber" : 24,
      "description" : "ErrorProne: Specify a `Locale` when calling `String#to{Lower,Upper}Case`. (Note: there are multiple suggested fixes; the third may be most appropriate if you're dealing with ASCII Strings.)    (see https://errorprone.info/bugpattern/StringCaseLocaleUsage)  Did you mean '.toLowerCase(Locale.ROOT);' or '.toLowerCase(Locale.getDefault());' or 'return Ascii.toLowerCase(currentSpan.getSpanContext()'?",
      "fingerprint" : "10542e731724aff6b495a26ed308eccebe18563fe10cf7aeca9320e5b6e70176",
      "fullPath" : "***MASKED***",
      "line" : 24,
      "relativePath" : "src/main/java/io/github/finoid/library/otel/OpenTelemetryUtils.java",
      "rule" : "StringCaseLocaleUsage",
      "severity" : "MINOR"
    }
  ]
]